application-admintools-ui/src/main/resources/AdminTools/Code/Macros.xml

```diff
@@ -152,5 +152,17 @@
     &lt;/div&gt;
   &lt;/div&gt;
 #end
+#macro (tableOfContentsUI)
+  &lt;div class="adminToolsDashboardHeader"&gt;
+    &lt;h1&gt;$escapetool.xml($services.localization.render('adminTools.dashboard.toc.title'))&lt;/h1&gt;
+    &lt;hr/&gt;
+    &lt;ul&gt;
+      &lt;li&gt;&lt;a href="#HBackend"&gt;$services.icon.renderHTML('world')
+        $escapetool.xml($services.localization.render('adminTools.dashboard.section.backend.title'))&lt;/a&gt;&lt;/li&gt;
+      &lt;li&gt;&lt;a href="#HFiles"&gt;$services.icon.renderHTML('download')
+        $escapetool.xml($services.localization.render('adminTools.dashboard.section.download.title'))&lt;/a&gt;&lt;/li&gt;
```

Comment:
Having a **static** table of contents poses some problems:
* it's hard to maintain: you need to remember to add / remove / update the entries when you modify the gadgets from the dashboard
* it can easily get out of sync: if I edit the dashboard and remove the Backend gadget for instance the table of contents still includes it, but I can't update the code (``AdminTools.Code.Macros``) because it is protected by the license

So overall I don't understand how a **static** table of contents can work on a dashboard that can be edited. Even if we make the default Admin Tools dashboard read-only, the admins will want to create their own, custom dashboard for sure, to re-arrange the gadgets or add / remove gadgets, so they won't use a static table of contents for sure.

Reply:
To make a dynamic table of content, I tried using the TOC macro, but I ran into some issues. 
* Firstly, I've noticed that the dashboard gadgets generate H1 headings even when the title field is left empty. Because of this, the TOC is displaying those empty headings as entries with the name of the page space, but no id associated: 
![image](https://github.com/xwikisas/application-admintools/assets/104432589/a3d35402-6a13-4a74-95af-8a4600119f05)

* The second issue I found is that the gadgets titles are missing an ID, which makes them unusable in a TOC macro. Even though the name of the title is correctly displayed in the TOC macro, because it is missing an ID it will point to the top of the page. 

The solution that I found is to add the titles inside the content of a gadget using XWiki syntax and render the TOC macro to start from level 2 headings. This way, we respect the idea of using H1 only for the page title, while also displaying the correct indentation inside the table of contents and avoiding the display of the gadget titles. The drawback is that the solution is only temporary, as when the dashboard will start using [H2 headings](https://jira.xwiki.org/browse/XWIKI-21370), they will be displayed too. 
To avoid this, I propose a function in JavaScript that removes those entries from the TOC macro that have reference to an empty ID. I found this solution while looking over an already closed issue: https://jira.xwiki.org/browse/XWIKI-20599 and the commit used to solve it: https://github.com/xwiki/xwiki-platform/commit/9a270960658c079edf684e3b08b42391e5bd4a98#diff-9e6ddce27ce20f417ba81fea56f38001c783b7b65c9d0b82ead6dfa6ffcb9466L275-L281. WDYT?
How it would look: 
![Screenshot from 2023-12-11 15-58-55](https://github.com/xwikisas/application-admintools/assets/104432589/86324188-8f88-4c21-b9fc-c1477feded80)

Also, regarding the problems described above, I found [this](https://jira.xwiki.org/browse/XWIKI-20600) opened issue for the second one, but no opened issue for the first one. Should I take another look and open an issue for the dashboard generating empty headings?

Reply:
> Should I take another look and open an issue for the dashboard generating empty headings?

I think you should open a jira issue for the first problem as well, yes.

Now, regarding the actual problem, for me your solution sounds ok. 
The problem is that when the dashboard headings will be of [H2 level](https://jira.xwiki.org/browse/XWIKI-21370) and have the [ID set](https://jira.xwiki.org/browse/XWIKI-20600), you will end up with a mixed behavior depending on the version where Admin tools is used if the issue about displayed empty headings is not solved, because you'll get back to the first problem described above. So maybe it's best to actually remove headings that don't have an id AND are empty, to be safe. Having this filter method actually removes also the need to starting with a specific heading level.
Users might still edit these gadgets and add their own titles, leading to 2 titled being displayed if all issues are fixed, but it sounds too far fetched to think about it, I'd say it's ok

Please also add a comprehensive comment about it in your code. Mention also the jira issues, since we might be able to use the gadget title once we will upgrade to a parent that has all these issues solved